iree/turbine/kernel/ops/wave_ops.py

```diff
@@ -1030,6 +1037,52 @@ def index(self, value: dict[IndexSymbol, IndexSequence]):
         CustomOp.index.fset(self, value)
 
 
+@define_op("extract")
+@dataclass
+class Extract(CustomOp):
+    """
+    Op Rationale:
+
+    Extract is an op used to represent extracting of
+    a scalar from TKW's 1-D vector on the specified index.
+
+    This can also be viewed as indexing/slicing on the fastest
+    dimension. Hence, the semantic of this op is designed to
+    see itself as a reduction on the indexed/fastest dimension.
```

Comment:
I am a little confused about how an extract is a reduction on the indexed/fastest dimension? Say you have a <8xf32> and you extract [0] you will get a f32, but how is this a reduction?

Reply:
Or maybe we should clarify - in the wave paradigm, when you do an extract of [M, N][0, :] -> [N] dimensional tensor, whereas in the thread paradigm, a <8xf32> -> f32 ? Is this what you mean?

Reply:
That is correct, say we have a tensor named `input` of shape [M, N], by indexing on the fastest dimension i.e `extracted = input[:, 3]` where 3 can be arbitary number between 0 and N. `extracted` will be of shape `Mx1` which is where the "reduction" came from.

Reply:
If so, I still would not call this a reduction.

Reply:
okay that makes sense. but I still wouldn't call it a reduction because that implies combining values in some way, whereas this is more of slicing.

Reply:
Going to resolve all these old comments for now.